File: clients/ios/Sources/Models/GbaStreamClient.swift
import Foundation

/// Thin Swift shell around the native session: connects, decodes, and
/// sends input entirely in C (unison_native_bridge.c + unison_core). This
/// class only owns the native handle and forwards calls -- direct port of
/// clients/android/.../GbaStreamClient.kt, with unison_native_bridge.c's
/// C function-pointer callbacks in place of JNI (see that file's own
/// comment for the full "no JNI-equivalent glue needed" rationale).
///
/// Like the Android original, Listener callbacks fire on the native
/// background thread, not the main thread -- same contract, same reason
/// (this class stays a thin, allocation-free-as-possible shell; hopping to
/// the main thread for UI updates is the caller's job, same as
/// PlayerActivity's own dispatch).
final class GbaStreamClient {
    protocol Listener: AnyObject {
        /// touchInput/hasButtons/hasSticks mirror GbaStreamClient.kt's
        /// Listener.onConnected -- see that interface's own extensive
        /// comment for the full input-encoding rationale. This MVP bridge
        /// only ever actually negotiates a gba_buttons (touchInput=false)
        /// session; the other fields exist for parity with a future
        /// touch-input PlayerView.
        /// streamType: the raw hello.stream_type ("GC_GBA_LINK",
        /// "N3DS_BOTTOM_SCREEN", ...) -- needed alongside
        /// touchInput/hasButtons/hasSticks because WIIU_GAMEPAD and
        /// N3DS_BOTTOM_SCREEN share the exact same true/true/true shape;
        /// only the raw string actually distinguishes Cemu from Azahar
        /// (used to key Prefs.bilinear(for:), which AntialiasingView's own
        /// rows set per stream type).
        func onConnected(touchInput: Bool, hasButtons: Bool, hasSticks: Bool, width: Int32, height: Int32,
                          grantedVideoMode: String, streamType: String)
        /// rgb565 is a fresh Data copy (unlike the C callback's raw
        /// pointer, which is only valid for that call) -- same "safe to
        /// keep" contract as Android's ByteArray.
        func onVideoFrame(width: Int32, height: Int32, rgb565: Data)
        /// UNISON_VIDEO_FORMAT_H264/_H265 only (mutually exclusive with
        /// onVideoFrame above) -- data is a raw Annex-B NAL stream, only
        /// valid for the duration of this call (unlike onVideoFrame's
        /// rgb565, NOT copied into a fresh Data first: a listener that
        /// needs to keep it past returning must copy it itself, since the
        /// whole point of this callback is to hand the bytes to
        /// CompressedVideoDecoder.decode(data:) synchronously without an
        /// extra copy on this per-frame hot path). width/height are the
        /// encoder's *coded* dimensions -- see unison_native_bridge.h's
        /// own comment on on_compressed_video_frame.
        func onCompressedVideoFrame(width: Int32, height: Int32, isH265: Bool, data: UnsafeRawBufferPointer)
        /// pcm is a fresh [Int16] copy, same reasoning as rgb565 above.
        func onAudioFrame(sampleRate: Int32, channels: Int32, pcm: [Int16])
        /// Server wants its own on-screen software keyboard shown (e.g.
        /// Cemu's WiiU swkbd) -- see unison_native_bridge.h's own comment
        /// on on_text_input_request for why the video stream never
        /// captures this itself. maxLength is 0 for "no limit enforced by
        /// the server". Reply with sendTextInputResponse() once the user
        /// submits or cancels, same GbaStreamClient.kt-mirrored contract.
        func onTextInputRequest(maxLength: Int32, initialText: String)
        func onDisconnected(reason: String)
    }

    static let videoModeDefault = Prefs.videoModeDefault
    // Mirrors GbaStreamClient.kt's own companion-object constants.
    static let streamTypeGcGbaLink = "GC_GBA_LINK"
    static let playerBasePort: Int32 = 6801

    private weak var listener: Listener?
    private var handle: OpaquePointer?
    // Keeps self alive across the C callback boundary (Unmanaged.passUnretained
    // in connect() only borrows a reference, it doesn't retain) -- released
    // in disconnect() once unison_native_disconnect() has joined the
    // background thread and guaranteed no further callback will fire.
    private var selfRetain: Unmanaged<GbaStreamClient>?

    init(listener: Listener) {
        self.listener = listener
    }

    /// Spawns a background native thread; connect result arrives via
    /// listener.onConnected/onDisconnected. videoMode is sent verbatim as
    /// hello_ack.video_mode (Prefs.videoMode(for:), one of Prefs.videoModes) --
    /// see docs/protocol.md; servers that don't implement the negotiation
    /// just ignore it.
    func connect(host: String, port: Int32, videoMode: String = GbaStreamClient.videoModeDefault) {
        let retained = Unmanaged.passRetained(self)
        selfRetain = retained

        var callbacks = unison_native_callbacks()
        callbacks.user_data = retained.toOpaque()
        callbacks.on_connected = { userData, touchInput, hasButtons, hasSticks, width, height, grantedVideoMode, streamType in
            guard let userData else { return }
            let client = Unmanaged<GbaStreamClient>.fromOpaque(userData).takeUnretainedValue()
            let mode = grantedVideoMode.map { String(cString: $0) } ?? ""
            let type = streamType.map { String(cString: $0) } ?? ""
            client.listener?.onConnected(touchInput: touchInput != 0, hasButtons: hasButtons != 0,
                                          hasSticks: hasSticks != 0, width: width, height: height,
                                          grantedVideoMode: mode, streamType: type)
        }
        callbacks.on_video_frame = { userData, width, height, rgb565, len in
            guard let userData, let rgb565 else { return }
            let client = Unmanaged<GbaStreamClient>.fromOpaque(userData).takeUnretainedValue()
            let data = Data(bytes: rgb565, count: len)
            client.listener?.onVideoFrame(width: width, height: height, rgb565: data)
        }
        callbacks.on_compressed_video_frame = { userData, width, height, isH265, data, len in
            guard let userData, let data else { return }
            let client = Unmanaged<GbaStreamClient>.fromOpaque(userData).takeUnretainedValue()
            let buffer = UnsafeRawBufferPointer(start: data, count: len)
            client.listener?.onCompressedVideoFrame(width: width, height: height, isH265: isH265 != 0, data: buffer)
        }
        callbacks.on_audio_frame = { userData, sampleRate, channels, pcm, sampleCount in
            guard let userData, let pcm else { return }
            let client = Unmanaged<GbaStreamClient>.fromOpaque(userData).takeUnretainedValue()
            let buffer = UnsafeBufferPointer(start: pcm, count: sampleCount)
            client.listener?.onAudioFrame(sampleRate: sampleRate, channels: channels, pcm: Array(buffer))
        }
        callbacks.on_text_input_request = { userData, maxLength, text, textLen in
            guard let userData else { return }
            let client = Unmanaged<GbaStreamClient>.fromOpaque(userData).takeUnretainedValue()
            // text isn't NUL-terminated (points straight at the WS payload
            // buffer, only valid for this call) -- String(cString:) would
            // assume NUL-termination and either read out of bounds or stop
            // early on an embedded NUL, so this decodes the explicit-length
            // buffer directly instead.
            let initialText: String
            if let text, textLen > 0 {
                let buffer = UnsafeRawBufferPointer(start: text, count: textLen)
                initialText = String(decoding: buffer, as: UTF8.self)
            } else {
                initialText = ""
            }
            client.listener?.onTextInputRequest(maxLength: Int32(truncatingIfNeeded: maxLength),
                                                 initialText: initialText)
        }
        callbacks.on_disconnected = { userData, reason in
            guard let userData else { return }
            let client = Unmanaged<GbaStreamClient>.fromOpaque(userData).takeUnretainedValue()
            let text = reason.map { String(cString: $0) } ?? ""
            client.listener?.onDisconnected(reason: text)
        }

        handle = unison_native_connect(host, port, videoMode, callbacks)
    }

    /// Sets the current gba_buttons keymask (GbaKey bits, OR'd together) --
    /// resent (only if changed) once per session-loop iteration, same
    /// "latest wins" contract as GbaStreamClient.kt's nativeSendInput.
    /// Only meaningful when listener.onConnected reported touchInput=false.
    func sendInput(keymask: UInt16) {
        unison_native_send_input(handle, keymask)
    }

    /// Sets the current touch state for a touch-capable session
    /// (onConnected reported touchInput=true) -- x/y are ignored whenever
    /// pressed is false, same convention as the C bridge and
    /// unison_touch_state itself. Which of the three wire shapes this
    /// actually sends is picked automatically by the C bridge from
    /// hasButtons/hasSticks, same as GbaStreamClient.kt's own
    /// nativeSendTouch.
    func sendTouch(pressed: Bool, x: UInt16, y: UInt16) {
        unison_native_send_touch(handle, pressed ? 1 : 0, x, y)
    }

    /// Sets the current buttons/analog-stick state -- only meaningful for
    /// a hasButtons session; see unison_native_send_extended_input's own
    /// header comment for why this doesn't itself trigger a send.
    func sendExtendedInput(buttons: UInt32, leftX: Int16, leftY: Int16, rightX: Int16, rightY: Int16) {
        unison_native_send_extended_input(handle, buttons, leftX, leftY, rightX, rightY)
    }

    /// Reply to listener.onTextInputRequest() -- confirmed=false (user
    /// cancelled) sends an empty text regardless of `text`, matching the C
    /// bridge's own convention. Only meaningful right after that callback
    /// fires, same one-shot contract as GbaStreamClient.kt's own
    /// sendTextInputResponse.
    func sendTextInputResponse(confirmed: Bool, text: String) {
        // `text` bridges to `const char *` automatically here, same as
        // host/videoMode already do in connect() above (a temporary
        // NUL-terminated UTF-8 buffer valid for the duration of this one
        // call) -- text.utf8.count is the explicit length the C side
        // actually uses (it doesn't rely on the NUL terminator).
        unison_native_send_text_input_response(handle, confirmed ? 1 : 0, text, text.utf8.count)
    }

    /// Blocks until the background thread has stopped (unison_native_
    /// disconnect() pthread_joins) -- guarantees no further Listener
    /// callback fires after this returns, same as the Kotlin original.
    func disconnect() {
        unison_native_disconnect(handle)
        handle = nil
        selfRetain?.release()
        selfRetain = nil
    }
}

File: clients/android/app/src/main/java/com/unison/android/GbaStreamClient.kt
package com.unison.android

/**
 * Thin Kotlin shell around the native session: connects, decodes, and sends
 * input entirely in C (jni_bridge.c + unison_core). This class only owns
 * the JNI handle and forwards calls; there's no protocol/transport logic on
 * the JVM side at all.
 */
class GbaStreamClient(private val listener: Listener) {

    interface Listener {
        // isTouch/hasButtons/hasSticks reflect the server's own
        // hello.input_encoding ("n3ds_touch_and_buttons" vs
        // "touch_and_buttons" vs "n3ds_touch" vs "gba_buttons", see
        // jni_bridge.c's perform_app_handshake), not a client-side guess
        // from the discovery beacon -- authoritative even for manual host
        // entry (which has no beacon to read a stream_type from beforehand)
        // and across a redirect hop landing on a different stream type than
        // the one first dialed. hasButtons is only ever true alongside
        // isTouch (there's no buttons-without-touch encoding); it's false
        // for "n3ds_touch" servers that don't accept remote buttons at all
        // (Cemu), true for both "touch_and_buttons" (melonDS's
        // NDS_BOTTOM_SCREEN -- buttons, but no analog sticks, the DS has
        // none on real hardware) and "n3ds_touch_and_buttons" (Azahar's
        // N3DS_BOTTOM_SCREEN -- buttons AND circle pad/analog sticks).
        // hasSticks distinguishes those last two; only ever true alongside
        // hasButtons.
        //
        // width/height are session_ready.video's final (possibly
        // downscaled) resolution -- known here regardless of which
        // video_mode ends up negotiated, unlike onVideoFrame's own
        // width/height (never called at all for h264/h265, see
        // PlayerActivity.onVideoFrame's own comment). This is what touch
        // input coordinate mapping should use.
        //
        // grantedVideoMode is session_ready.video_mode verbatim -- empty if
        // the server predates that field entirely (see docs/protocol.md
        // "Video-mode fallback"). Compare against Prefs.videoModeFor() (what
        // was actually requested, see connect()'s own videoMode param) to
        // decide whether to prompt: skip the comparison entirely if this is
        // blank, don't treat blank as "tiles was granted".
        fun onConnected(isTouch: Boolean, hasButtons: Boolean, hasSticks: Boolean, width: Int, height: Int, grantedVideoMode: String)
        fun onVideoFrame(width: Int, height: Int, rgb565: ByteArray)
        fun onAudioFrame(sampleRate: Int, channels: Int, pcm: ShortArray)
        // The server's own on-screen software keyboard (e.g. Cemu's swkbd)
        // is drawn as a host-side UI overlay, never part of the captured
        // video -- this is the server asking the client to show its own
        // native text input UI instead, since there'd otherwise be no way
        // to type at all. maxLength is in characters, 0 = no server-side
        // limit; initialText is whatever's already in the field (often
        // empty). Reply with sendTextInputResponse() once the user submits
        // or cancels.
        fun onTextInputRequest(maxLength: Int, initialText: String)
        // Mirrors real mic hardware: the console only wants microphone
        // input while a game has it powered on and actively sampling (see
        // e.g. the 3DS's mic:u service or the Wii U GamePad's own mic
        // input), not continuously just because a stream is connected --
        // start (or stop) capturing from the device's own microphone via
        // sendMicAudio() only while enabled is true, at sampleRate (the
        // exact rate the console asked for; AudioRecord accepts arbitrary
        // rates and resamples internally, so this can be requested
        // directly without the client doing its own resampling). Requires
        // RECORD_AUDIO at runtime -- if the permission isn't granted, the
        // implementation should simply not capture rather than crash.
        fun onMicEnable(enabled: Boolean, sampleRate: Int)
        fun onDisconnected(reason: String)
    }

    @Volatile
    private var nativeHandle: Long = 0

    /** Spawns a background native thread; connect result arrives via onConnected/onDisconnected.
     * videoMode is sent verbatim as hello_ack.video_mode (Prefs.videoModeFor(streamType), one of
     * Prefs.VIDEO_MODES) -- see docs/protocol.md; servers that don't implement the negotiation
     * just ignore it. */
    fun connect(host: String, port: Int, videoMode: String = Prefs.VIDEO_MODE_DEFAULT) {
        nativeHandle = nativeConnect(host, port, videoMode, listener)
    }

    /** Hands the TextureView's Surface (see PlayerScreen's video layer) down
     * to native code, which wraps it as an ANativeWindow for MediaCodec's
     * Surface-mode output -- only meaningful for UNISON_VIDEO_FORMAT_H264/
     * _H265 frames (see jni_bridge.c's ensure_video_codec()); harmless,
     * just unused, for tiles/legacy sessions. Pass null when the surface is
     * destroyed (e.g. TextureView.SurfaceTextureListener.onSurfaceTextureDestroyed). */
    fun setVideoSurface(surface: android.view.Surface?) {
        val handle = nativeHandle
        if (handle != 0L) nativeSetVideoSurface(handle, surface)
    }

    /** Cheap: just records the latest key state, the native session loop sends it.
     * Only meaningful for a gba_buttons session (Listener.onConnected(isTouch = false)). */
    fun sendInput(keyMask: Int) {
        val handle = nativeHandle
        if (handle != 0L) nativeSendInput(handle, keyMask)
    }

    /** Touch counterpart to sendInput -- only meaningful for a touch session
     * (Listener.onConnected(isTouch = true)). x/y are in the current video
     * frame's own native pixel coordinates (see PlayerActivity's touch
     * overlay for the tap-position mapping); ignored by the receiver
     * whenever pressed is false (unison_touch_state's own convention,
     * protocol.h), so any value is fine there -- 0 is simplest. */
    fun sendTouch(pressed: Boolean, x: Int, y: Int) {
        val handle = nativeHandle
        if (handle != 0L) nativeSendTouch(handle, pressed, x, y)
    }

    /** Combined counterpart to sendTouch -- only meaningful for a session
     * where Listener.onConnected reported hasButtons = true. buttons is an
     * OR of the BUTTON_* constants below; leftX/leftY is the circle pad or,
     * on a two-stick console, the left stick (-32768..32767 per axis, 0
     * centered); rightX/rightY is always 0 from a caller with only one
     * stick to report. touchPressed/touchX/touchY follow sendTouch's own
     * convention (x/y ignored by the receiver whenever touchPressed is
     * false). Stick args are meaningless (and simply dropped native-side,
     * see jni_bridge.c's maybe_send_touch()) when hasSticks = false --
     * callers there (no VirtualStick UI shown at all) should just always
     * pass 0. */
    fun sendExtendedInput(
        touchPressed: Boolean, touchX: Int, touchY: Int,
        buttons: Int, leftX: Int, leftY: Int, rightX: Int = 0, rightY: Int = 0
    ) {
        val handle = nativeHandle
        if (handle != 0L) {
            nativeSendExtendedInput(handle, touchPressed, touchX, touchY, buttons, leftX, leftY, rightX, rightY)
        }
    }

    /** Reply to Listener.onTextInputRequest() -- confirmed=false (user
     * cancelled) sends regardless of what's in text, the server is expected
     * to leave its existing text unchanged in that case (Unison's own
     * unison_text_input_response convention). */
    fun sendTextInputResponse(confirmed: Boolean, text: String) {
        val handle = nativeHandle
        if (handle != 0L) nativeSendTextInputResponse(handle, confirmed, text)
    }

    /** Uploads a chunk of mono s16 PCM samples captured from the device's own
     * microphone -- only meaningful while Listener.onMicEnable(true, ...)
     * is the most recent state; harmless (silently queued, just never
     * useful) otherwise. sampleRate should match whatever onMicEnable most
     * recently reported. */
    fun sendMicAudio(sampleRate: Int, samples: ShortArray) {
        val handle = nativeHandle
        if (handle != 0L) nativeSendMicAudio(handle, sampleRate, samples)
    }

    fun disconnect() {
        val handle = nativeHandle
        if (handle != 0L) {
            nativeHandle = 0
            nativeDisconnect(handle)
        }
    }

    private external fun nativeConnect(host: String, port: Int, videoMode: String, listener: Listener): Long
    private external fun nativeSetVideoSurface(handle: Long, surface: android.view.Surface?)
    private external fun nativeSendInput(handle: Long, keyMask: Int)
    private external fun nativeSendTouch(handle: Long, pressed: Boolean, x: Int, y: Int)
    private external fun nativeSendExtendedInput(
        handle: Long, touchPressed: Boolean, touchX: Int, touchY: Int,
        buttons: Int, leftX: Int, leftY: Int, rightX: Int, rightY: Int
    )
    private external fun nativeSendTextInputResponse(handle: Long, confirmed: Boolean, text: String)
    private external fun nativeSendMicAudio(handle: Long, sampleRate: Int, samples: ShortArray)
    private external fun nativeDisconnect(handle: Long)

    companion object {
        // Player ports 6801-6804 (docs/protocol.md); /status on each one
        // (not a combined endpoint) is still how the P1-P4 picker below
        // finds a free slot -- the app-level handshake (core's
        // unison/handshake.h, spoken over the WebSocket connection itself
        // once nativeConnect() dials a specific port) is a separate,
        // later step, not a replacement for this pre-connect check.
        const val PLAYER_BASE_PORT = 6801
        const val PLAYER_SLOT_COUNT = 4

        // The one stream type with a lobby port fanning out to separate
        // per-slot ports (see the P1-P4 picker above) -- every other stream
        // type (N3DS_BOTTOM_SCREEN, NDS_BOTTOM_SCREEN, WIIU_GAMEPAD, ...) is
        // single-client and its beacon's handshake_port *is* the only port
        // there is, so MenuActivity connects to it directly instead of
        // running the GC_GBA_LINK-specific slot probe against it (which
        // used to always report "alle Plätze belegt" for these -- ports
        // 6801-6804 never existed on a server that isn't Dolphin).
        const val STREAM_TYPE_GC_GBA_LINK = "GC_GBA_LINK"

        // Mirrors UNISON_PROTOCOL_VERSION (core/include/unison/handshake.h)
        // and UNISON_BEACON_PORT (core/include/unison/discovery.h) --
        // MenuActivity's discovery listener needs both before any native
        // handshake code runs, so it can't just call into core for them.
        const val PROTOCOL_VERSION = 2
        const val BEACON_PORT = 6805

        // Mirrors unison_key in core/include/unison/protocol.h.
        const val KEY_A = 1 shl 0
        const val KEY_B = 1 shl 1
        const val KEY_SELECT = 1 shl 2
        const val KEY_START = 1 shl 3
        const val KEY_RIGHT = 1 shl 4
        const val KEY_LEFT = 1 shl 5
        const val KEY_UP = 1 shl 6
        const val KEY_DOWN = 1 shl 7
        const val KEY_R = 1 shl 8
        const val KEY_L = 1 shl 9

        // Mirrors unison_button_bit in core/include/unison/protocol.h --
        // for sendExtendedInput's buttons parameter, not sendInput's
        // keyMask (KEY_* above, an unrelated bit layout for gba_buttons).
        // A given bit only means something to a server whose console
        // actually has that button; see unison_button_bit's own comment.
        const val BUTTON_A = 1 shl 0
        const val BUTTON_B = 1 shl 1
        const val BUTTON_X = 1 shl 2
        const val BUTTON_Y = 1 shl 3
        const val BUTTON_L = 1 shl 4
        const val BUTTON_R = 1 shl 5
        const val BUTTON_ZL = 1 shl 6
        const val BUTTON_ZR = 1 shl 7
        const val BUTTON_SELECT = 1 shl 8 // aka Minus (Wii U)
        const val BUTTON_START = 1 shl 9  // aka Plus (Wii U)
        const val BUTTON_UP = 1 shl 10
        const val BUTTON_DOWN = 1 shl 11
        const val BUTTON_LEFT = 1 shl 12
        const val BUTTON_RIGHT = 1 shl 13
        const val BUTTON_HOME = 1 shl 14

        init {
            System.loadLibrary("unison_android")
        }
    }
}
